lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -507,6 +508,25 @@ LogicalResult BuildOpGroups::buildOp(PatternRewriter &rewriter,
   return success();
 }
 
+LogicalResult BuildOpGroups::buildOp(PatternRewriter &rewriter,
+                                     memref::CopyOp copyOp) const {
+  auto source = copyOp.getSource();
+  auto target = copyOp.getTarget();
+
+  auto srcType = source.getType().cast<MemRefType>();
+  auto tgtType = target.getType().cast<MemRefType>();
+
+  assert(srcType.getElementType() == tgtType.getElementType() &&
+         srcType.getShape() == tgtType.getShape() &&
+         "Source and target are expected to have the same element type and "
+         "shape.");
+
+  copyOp.emitError("Please run --handshake-legalize-memrefs to replace "
```

Comment:
I'm all for good error messages, but it seems a little weird to handle `memref::CopyOp` only to have `buildOp` unconditionally throw an error.